src/main/scala/lc/Main.scala

```diff
@@ -163,18 +173,72 @@ case class Id(id: String)
 case class Answer(answer: String, id: String)
 case class Secret(token: Int)
 
+class RateLimiter extends DBConn {
+  val stmt = getConn()
+  val userLastActive = collection.mutable.Map[Int, Long]()
+  val userAllowance = collection.mutable.Map[Int, Double]() 
+  val rate = 2.0 
+  val per = 45.0
+  val allowance = rate
+
+  def validateUser(user: Int) : Boolean = {
+    val allow = if(userLastActive.contains(user)){
+      true
+    } else {
+      validatePstmt.setInt(1, user)
+      val rs = validatePstmt.executeQuery()
+      val validated = if(rs.next()){
+        val hash = rs.getInt("hash")
+        userLastActive(hash) = System.currentTimeMillis()
+        userAllowance(hash) = allowance
+        true
+      } else {
+        false
+      }
+      validated
+    }
+    allow
+  }
+
+  def checkLimit(user: Int): Boolean = {
+    synchronized {
+      val current = System.currentTimeMillis()
+      val time_passed = (current - userLastActive(user)) / 1000000000 
+      userLastActive(user) = current
+      userAllowance(user) += time_passed * (rate/per)
+      if(userAllowance(user) > rate){ userAllowance(user) = rate }
+      val allow = if(userAllowance(user) < 1.0){
+        false
+      } else {
+        userAllowance(user) -= 1.0
+        true
+      }
+      allow
+    }
+  }
+
+}
+
 class Server(port: Int){
 	val captcha = new Captcha(0)
+  val rateLimiter = new RateLimiter()
 	val server = new HTTPServer(port)
 	val host = server.getVirtualHost(null)
 
 	implicit val formats = DefaultFormats
 
 	host.addContext("/v1/captcha",(req, resp) => {
-    	val body = req.getJson()
-    	val json = parse(body)
-    	val param = json.extract[Parameters]
-    	val id = captcha.getChallenge(param)
+      val accessToken = if(req.getHeaders().get("access-token") != null){
+        req.getHeaders().get("access-token").toInt
+      } else 0
```

Comment:
You can read up on scala's `Option` type. It can simplify code a bit, for example:
```scala
//  Here the type of accessToken will be Option[String]
val accessToken = Option(req.getHeaders().get("access-token"))   

//  Here the type of accessToken will be Option[Int]
val accessToken = Option(req.getHeaders().get("access-token")).map(_.toInt)   

val access = accessToken.map(t => rateLimiter.validateUser(t) && rateLimite.checkLimit(t)).getOrElse(false)

val id = if (access) {
  val body = req.getJson()
  val json = parse(body)
  val param = json.extract[Parameters]
  captcha.getChallenge(param)
} else {
  "Not a valid user or rate limit reached!"
}
```

Reply:
Yes, this looks good.